File: MATLAB/ParallelPlateCapacitorFEM.m
clear
clc

w = 4;
h = 0.2;
d = 1;
er = 2.2;
V = 100;

A = 2*w;
B = 2*w;

% Geometry Description
gd = [3 3 3; 
      4 4 4; 
      -A/2 w/2 w/2; 
      A/2 w/2 w/2; 
      A/2 -w/2 -w/2; 
      -A/2 -w/2 -w/2;
      -B/2 d/2 -d/2; 
      -B/2 d/2+h -d/2-h; 
      B/2 d/2+h -d/2-h; 
      B/2 d/2 -d/2];
sf = 'R1-R2-R3';
ns = [82 82 82; 49 50 51];
dl = decsg(gd, sf, ns);

% Create Mesh
[p, e, t] = initmesh(dl);
[p, e, t] = refinemesh(dl, p, e, t);
[p, e, t] = refinemesh(dl, p, e, t);
[p, e, t] = refinemesh(dl, p, e, t);
Num_nodes = size(p, 2);
Num_elements = size(t, 2);


% Define known and unknown potentials
node_id = ones(Num_nodes, 1);
Potentials = zeros(Num_nodes, 1);
Num_edges = size(e, 2);
for ie = 1:Num_edges
    region1 = e(6, ie);
    region2 = e(7, ie);
    nodes(1:2) = e(1:2, ie);
    x(1:2) = p(1, nodes(1:2));
    y(1:2) = p(2, nodes(1:2));

    if (region1 == 0 || region2 == 0)
        if  (y(1) >= d/2 - h) && (y(1) <= d/2 + h) && (x(1) >= -w/2) && (x(1) <= w/2)
            Potentials(nodes(1)) = V/2;
            node_id(nodes(1)) = 0;
        elseif (y(1) >= -d/2 - h) && (y(1) <= -d/2 + h) && (x(1) >= -w/2) && (x(1) <= w/2)
            Potentials(nodes(1)) = -V/2;
            node_id(nodes(1)) = 0;
        end
        if (y(2) >= d/2 - h) && (y(2) <= d/2 + h) && (x(2) >= -w/2) && (x(2) <= w/2)
            Potentials(nodes(1)) = V/2;
            node_id(nodes(1)) = 0;
        elseif (y(2) >= -d/2 - h) && (y(2) <= -d/2 + h) && (x(2) >= -w/2) && (x(2) <= w/2)
            Potentials(nodes(1)) = -V/2;
            node_id(nodes(1)) = 0;
        end
    end
end

% Indexes of unknowns
non_pot_index = zeros(Num_nodes, 1);
counter = 0;
for in = 1:Num_nodes
    if (node_id(in) == 1)
        counter = counter + 1;
        non_pot_index(in) = counter;
    end
end

% Matrix Calculation
Nf = counter;
Sff = spalloc(Nf, Nf, 7*Nf);
B = zeros(Nf, 1);

for ie = 1:Num_elements
    nodes(1:3) = t(1:3, ie);
    region = t(4, ie);
    x(1:3) = p(1, nodes(1:3));
    y(1:3) = p(2, nodes(1:3));

    De = det([1 x(1) y(1); 1 x(2) y(2); 1 x(3) y(3)]);
    Ae = abs(De/2);
    b(1) = (y(2) - y(3))/De;
    b(2) = (y(3) - y(1))/De;
    b(3) = (y(1) - y(2))/De;
    c(1) = (x(3) - x(2))/De;
    c(2) = (x(1) - x(3))/De;
    c(3) = (x(2) - x(1))/De;

    for i = 1:3
        for j = 1:3
            flag(1) = (x(i) >= -w/2 && x(i) <= w/2) && (y(i) >= -d/2 && y(i) <= d/2);
            flag(2) = (x(j) >= -w/2 && x(j) <= w/2) && (y(j) >= -d/2 && y(j) <= d/2);
            if(flag(1) && flag(2))
                Se(i,j) = er*(b(i)*b(j) + c(i)*c(j))*Ae;
            else
                Se(i, j) = (b(i)*b(j) + c(i)*c(j))*Ae;
            end

            if (node_id(nodes(i)) == 1)
                if (node_id(nodes(j)) == 1)
                    Sff(non_pot_index(nodes(i)), non_pot_index(nodes(j))) = ...
                    Sff(non_pot_index(nodes(i)), non_pot_index(nodes(j))) + Se(i, j);
                else
                    B(non_pot_index(nodes(i))) = B(non_pot_index(nodes(i))) - ...
                    Se(i, j)*Potentials(nodes(j));
                end
            end
        end
    end
end

% Unknown Potential Calculation 
pot = Sff\B;
for in = 1:Num_nodes
    if node_id(in) == 1
        Potentials(in) = pot(non_pot_index(in));
    end
end

% Calculate Electric Field
[Ex, Ey] = pdegrad(p, t, Potentials);
Ex = -Ex;
Ey = -Ey;

% Plot the Results
figure(1);
pdeplot(p, e, t, 'xydata', Potentials, 'contour', 'on');
axis tight;
axis equal;
hold on;
pdeplot(p, e, t, 'FlowData', [Ex', Ey']');
colormap jet;

File: Python/ParallelPlateCapacitorFEM.py
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.tri as mtri
import scipy.sparse as sp

w = 4
t = 0.2
d = 1
V = 100
er = 2.2

A = 2*w
B = w


# Create Geometry
x_n = np.linspace(-A/2, A/2, 100)
y_n = np.linspace(-B/2, B/2, 100)
X_n, Y_n = np.meshgrid(x_n, y_n)

X = X_n.reshape(X_n.size)
Y = Y_n.reshape(Y_n.size)

triang = mtri.Triangulation(X, Y)

tri_coords = np.zeros((triang.triangles.shape[0], 3, 2))
for i, tri in enumerate(triang.triangles):
    tri_coords[i] = np.column_stack((X[tri], Y[tri]))

mask = np.zeros(triang.triangles.shape[0], dtype=bool)
for i in range(0, triang.triangles.shape[0]):
    t_c = tri_coords[i]
    x = [t_c[0][0], t_c[1][0], t_c[2][0]]
    y = [t_c[0][1], t_c[1][1], t_c[2][1]]
    # Capacitor Plates:
    # x: [-w/2, w/2]
    # y: [-d/2-t, -d/2] and [d/2, d/2+t]
    flag = np.zeros(3, dtype = 'bool')
    for j in range(0, 3):
        flag[j] = (x[j] >= -w/2 and x[j] <= w/2) and ((y[j] >= -d/2-t and y[j] <= -d/2) or (y[j] >= d/2 and y[j] <= d/2 + t))
    if flag[0] and flag[1] and flag[2]:
        mask[i] = True
triang.set_mask(mask)
nodes = np.column_stack((triang.x, triang.y))
unique_nodes = np.unique(nodes, axis=0)
element_nodes = triang.triangles


# Define known and unknown potentials
node_id = np.ones(nodes.shape[0])
Potentials = np.zeros(nodes.shape[0])
for inode in range(0, nodes.shape[0]):
    x = nodes[inode][0]
    y = nodes[inode][1]
    if (x >= -w/2 - pow(10,-6) and x <= w/2 + pow(10,-6)):
        if (y >= -d/2- t - pow(10,-6) and y <= -d/2 + pow(10,-6)):
            node_id[inode] = 0
            Potentials[inode] = -V/2
        elif (y >= d/2 - pow(10,-6) and y <= d/2 + t + pow(10,-6)):
            node_id[inode] = 0
            Potentials[inode] = V/2

# Index of unkowns
un_index = np.zeros(nodes.size, dtype=int)
counter = 0
for inode in range(0, nodes.shape[0]):
    if node_id[inode] == 1:
        un_index[inode] = counter
        counter = counter + 1


# Matrix Calculation
Nf = counter
Sff = sp.lil_matrix((Nf, Nf), dtype = float)
B = np.zeros(Nf)

for ie in range(0, triang.triangles.shape[0]):
    t_c = tri_coords[ie]
    x = np.zeros(3)
    y = np.zeros(3)
    x = [t_c[0][0], t_c[1][0], t_c[2][0]]
    y = [t_c[0][1], t_c[1][1], t_c[2][1]]
    n = [element_nodes[ie][0], element_nodes[ie][1], element_nodes[ie][2]]
        
    A = np.array([[1, x[0], y[0]], [1, x[1], y[1]], [1, x[2], y[2]]])
    De = np.linalg.det(A)
    Ae = abs(De/2)
    b = np.zeros(3)
    c = np.zeros(3)
    b = [(y[1]-y[2])/De, (y[2]-y[0])/De, (y[0]-y[1])/De]
    c = [(x[2]-x[1])/De, (x[0]-x[2])/De, (x[1]-x[0])/De]

    Se = np.zeros([3, 3])
    for i in range(0, 3):
        for j in range(0, 3):
            flag[1] = (x[i] >= -w/2 and x[i] <= w/2) and (y[i] >= -d/2 and y[i] <= d/2 )
            flag[2] = (x[j] >= -w/2 and x[j] <= w/2) and (y[j] >= -d/2 and y[j] <= d/2 )
            if flag[1] and flag[2]:
                Se[i][j] = er*(b[i]*b[j] + c[i]*c[j])*Ae
            else:
                Se[i][j] = (b[i]*b[j] + c[i]*c[j])*Ae

            if node_id[n[i]] == 1:
                if node_id[n[j]] == 1:
                    Sff[un_index[n[i]], un_index[n[j]]] = Sff[un_index[n[i]], un_index[n[j]]] + Se[i][j]
                else:
                    B[un_index[n[i]]] = B[un_index[n[i]]] - Se[i][j]*Potentials[n[j]]


# Unkown Potential Calculation
P = sp.linalg.bicgstab(Sff, B)
pot = P[0]
for inode in range(0, nodes.shape[0]):
    if node_id[inode] == 1:
        Potentials[inode] = pot[un_index[inode]]


# Electric Field Calculation
interpolator = mtri.LinearTriInterpolator(triang, Potentials)
[Ex, Ey] = interpolator.gradient(triang.x, triang.y)


 #Plot the results
fig, ax = plt.subplots(1)
ax.triplot(triang)
cax = ax.tripcolor(triang, Potentials, cmap='plasma', shading='flat')
fig.colorbar(cax, ax=ax)
ax.set_xlabel('x')
ax.set_ylabel('y')
plt.show()

